### PR TITLE
[CSApply] Don't diagnose a 'static let none' as ambiguous none unless its an enum type

### PR DESCRIPTION
**Pull request by theblixguy**
We were incorrectly diagnosing the following:

```swift
enum Foo {
  case bar
  static let none = 1
}

let optionalFoo: Foo? = .none // warning: assuming you mean 'Optional<Foo>.none'; 
                              // did you mean 'Foo.none' instead?
```

There's no ambiguity here because `.none` could only refer to `Optional.none`, since the type of `Foo.none` is `Int`.

We should still offer a warning in case the user does something like this:

```swift
enum Foo {
  case bar
  static let none = Foo.bar
}

let optionalFoo: Foo? = .none // warning
```

Resolves [SR-11522](https://bugs.swift.org/browse/SR-11522).

### CODE REVIEW

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -2487,7 +2487,14 @@ namespace {
...
             }
-            
+
+            // If we have something like 'static let none = 1', then ignore.
+            if (auto VD = dyn_cast<VarDecl>(member)) {
+              if (!VD->getInterfaceType()->isEqual(
+                      baseTyNominalDecl->getDeclaredInterfaceType()))
+                return;
+            }
+
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: While you're here, what happens if there's a static *method* named `none`? Seems like that's similar to the enum-with-associated-values case.

Comment (jrose-apple) [reply lib/Sema/CSApply.cpp]: And what if there's both an instance member and a static member named `none`?

File: test/decl/enum/enumtest.swift [hunk 2]
```diff
@@ -441,4 +441,21 @@ class C {}
...
+
+enum EnumWithStaticNone2 {
+  case a
+  static let none = EnumWithStaticNone2.a
+}
+
+let _: EnumWithStaticNone1? = .none // Okay
+let _: EnumWithStaticNone2? = .none // expected-warning {{assuming you mean 'Optional<EnumWithStaticNone2>.none'; did you mean 'EnumWithStaticNone2.none' instead?}}
+// expected-note@-1 {{explicitly specify 'Optional' to silence this warning}}{{31-31=Optional}}
+// expected-note@-2 {{use 'EnumWithStaticNone2.none' instead}}{{31-31=EnumWithStaticNone2}}
```
Comment (brentdax) [inline test/decl/enum/enumtest.swift]: How do these test cases behave? How *should* they behave?

```swift
enum GenericEnumWithStaticNone<T> {
  case a
  static var none: GenericEnumWithStaticNone<Int> { .a }
}

let _: GenericEnumWithStaticNone<Int>? = .none
let _: GenericEnumWithStaticNone<String>? = .none
let _: GenericEnumWithStaticNone? = .none
```

**Review by beccadax — CHANGES_REQUESTED** [review]
Good idea!

File: test/decl/enum/enumtest.swift (hunk 2 above)
Comment (jrose-apple) [reply test/decl/enum/enumtest.swift]: I don't think this should necessarily block *this* PR, though.

Comment (jrose-apple) [reply test/decl/enum/enumtest.swift]: Another variation:

```swift
enum GenericEnum<T> {
  case a
}
extension GenericEnum where T == Int {
  static var none: GenericEnumWithStaticNone<Int> { .a }
}
```

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: We don't include instance members in the check, only static members. I suppose for the method case, we can check the result type of the method:

```swift
enum Foo {
  case bar
  static func none() -> Foo { return Foo.bar }
}

let _: Foo? = .none // warning
```

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: Obviously, the warning has to be tweaked, so it mentions calling `Foo.none()`

File: test/decl/enum/enumtest.swift (hunk 2 above)
Comment (theblixguy) [reply test/decl/enum/enumtest.swift]: None of them produce a warning at the moment.

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/Sema/CSApply.cpp]: > We don't include instance members in the check

The current implementation early-exits if it sees an instance member, even if there's also a static member. That's why I brought it up.

I'm okay with the method case just being skipped too; the point is that it's *not* ambiguous (though as Brent points out the criterion is "confusing" more than "ambiguous").

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: Oh, we already diagnose (even without this patch):

```
enum Foo {
  case bar
  static let none = Foo.bar
  var none: Foo { return .bar }
}

let optionalFoo: Foo? = .none // warning
```

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: Ah of course, it depends on source order 😅 That should be fixed!

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: Fixed.

File: test/decl/enum/enumtest.swift (hunk 2 above)
Comment (theblixguy) [reply test/decl/enum/enumtest.swift]: I think the first one should produce a warning, but the latter two shouldn't since `none`'s type is `GenericEnumWithStaticNone<Int>` but only first one has it as the base type.

@jrose-apple Did you mean `static var none: GenericEnum <Int> { .a }`?

Comment (theblixguy) [reply test/decl/enum/enumtest.swift]:
```swift
enum GenericEnumWithStaticNone<T> {
  case a
  static var none: GenericEnumWithStaticNone<Int> { .a }
}

let _: GenericEnumWithStaticNone<Int>? = .none // warning
let _: GenericEnumWithStaticNone<String>? = .none // okay
let _: GenericEnumWithStaticNone? = .none // okay

enum GenericEnum<T> {
  case a
}

extension GenericEnum where T == Int {
  static var none: GenericEnum<Int> { .a }
}

let _: GenericEnum<Int>? = .none // warning
let _: GenericEnum<String>? = .none // okay
```

Comment (theblixguy) [reply test/decl/enum/enumtest.swift]: Alright, should be diagnosed now.

File: test/decl/enum/enumtest.swift [hunk 3]
```diff
@@ -441,4 +441,85 @@ class C {}
...
+enum GenericEnumWithStaticNone<T> {
+  case a
+  static var none: GenericEnumWithStaticNone<Int> { .a }
+}
+
+let _: GenericEnumWithStaticNone<Int>? = .none // expected-warning {{assuming you mean 'Optional<GenericEnumWithStaticNone<Int>>.none'; did you mean 'GenericEnumWithStaticNone<Int>.none' instead?}}
+// expected-note@-1 {{explicitly specify 'Optional' to silence this warning}}{{42-42=Optional}}
+// expected-note@-2 {{use 'GenericEnumWithStaticNone<Int>.none' instead}}{{42-42=GenericEnumWithStaticNone<Int>}}
+let _: GenericEnumWithStaticNone<String>? = .none // Okay
+let _: GenericEnumWithStaticNone? = .none // Okay
```
Comment (brentdax) [inline test/decl/enum/enumtest.swift]: Huh, that's interesting. I think this doesn't warn because solutions that select `Optional.none` don't pin down `T`, so they're rejected as ambiguous.

I think we'd like it to warn about this case, but I'm guessing that isn't something you could fix easily. Maybe file a bug about it and add a FIXME with the bug number?

Comment (theblixguy) [reply test/decl/enum/enumtest.swift]: In this case, we're not assigning `Optional.none`, but rather `GenericEnumWithStaticNone<Int>.none` so we don't even enter the check.

File: lib/Sema/CSApply.cpp [hunk 4]
```diff
@@ -2475,25 +2475,34 @@ namespace {
...
-            
-            // Return if the member is an enum case w/ assoc values, as we only
-            // care (for now) about cases with no assoc values (like none)
-            if (auto EED = dyn_cast<EnumElementDecl>(member)) {
-              if (EED->hasAssociatedValues()) {
+            // If this is a variable whose type does not match the base type
+            // of the target then ignore it.
+            if (auto VD = dyn_cast<VarDecl>(member)) {
+              auto baseType = DSCE->getType()->lookThroughAllOptionalTypes();
+              if (!VD->getInterfaceType()->isEqual(baseType))
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: Why is this one checked here but the other tests done in the filter?

(I was just going to suggest looping over the results and using `continue` instead of `return`, but filtering works too.)

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: No good reason I guess, let me move it into the filter.

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: Done

File: test/decl/enum/enumtest.swift (hunk 3 above)
Comment (theblixguy) [reply test/decl/enum/enumtest.swift]: Added a FIXME.

**Comment by theblixguy** [comment]
Alright, this should be good to go now.

@swift-ci please smoke test

File: lib/Sema/CSApply.cpp [hunk 5]
```diff
@@ -2475,28 +2475,34 @@ namespace {
...
+              if (auto EED = dyn_cast<EnumElementDecl>(member)) {
+                return !EED->hasAssociatedValues();
+              }
+              if (auto VD = dyn_cast<VarDecl>(member)) {
+                auto baseType = DSCE->getType()->lookThroughAllOptionalTypes();
+                return VD->getInterfaceType()->isEqual(baseType);
+              }
+            }
+
+            return true;
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: Do we think there'll be any other results, or should this be `llvm_unreachable`?

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: We might have a `typealias none = <whatever>` :-)

Comment (jrose-apple) [reply lib/Sema/CSApply.cpp]: True. If one of those is there, what would the right thing to do be?

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: Hmm I thought we'd get a redeclaration error, but seems like we don't get one if you have a computed property. So, I think the right thing to do here is to simply filter them out, otherwise we're gonna trip the warning:

```swift
enum Foo {
  case bar
  var none: Foo { . bar }
  struct none {}
}

let foo: Foo? = .none // warning 
```

Comment (jrose-apple) [reply lib/Sema/CSApply.cpp]: That's an instance property rather than a static property. I was thinking of the case where you *just* have the type, though.

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: If we just have `struct none {}` then we'd trip the warning as well. Anyway, since it doesn't happen now because it gets filtered out.

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: I have added some more tests (it's getting really edge case-y now 😅).

File: lib/Sema/CSApply.cpp [hunk 6]
```diff
@@ -2475,28 +2475,37 @@ namespace {
...
+              if (auto VD = dyn_cast<VarDecl>(member)) {
+                auto baseType = DSCE->getType()->lookThroughAllOptionalTypes();
+                return VD->getInterfaceType()->isEqual(baseType);
+              }
+            }
+
+            // Filter out anything that's not one of the above. We don't care
+            // if we have a typealias named 'none' or a struct/class named
+            // 'none'.
+            return false;
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: This makes more sense to me, thanks. :-)

File: test/decl/enum/enumtest.swift [hunk 7]
```diff
@@ -441,4 +441,114 @@ class C {}
...
+enum EnumWithBothTypealiasAndComputedNone {
+  case bar
+  typealias none = EnumWithBothTypealiasAndComputedNone
+  var none: EnumWithBothTypealiasAndComputedNone { . bar }
+}
+
+let _: EnumWithStructNone? = .none // Okay
+let _: EnumWithTypealiasNone? = .none // Okay
+let _: EnumWithBothStructAndComputedNone? = .none // Okay
+let _: EnumWithBothTypealiasAndComputedNone? = .none // Okay
```
Comment (jrose-apple) [inline test/decl/enum/enumtest.swift]: It's good to have these test cases in the compiler. :-)

**Review by jrose-apple — APPROVED** [review]

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Review by beccadax — APPROVED** [review]

**Comment by theblixguy** [comment]
@swift-ci please smoke test Linux platform

**Comment by theblixguy** [comment]
@swift-ci please smoke test Linux platform
